gazebo_ros_control/include/gazebo_ros_control/default_robot_hw_sim.h

```diff
@@ -0,0 +1,173 @@
+#include <gazebo/gazebo.hh>
+
+// ROS
+#include <rclcpp/rclcpp.hpp>
+#include <angles/angles.h>
+#include <pluginlib/class_list_macros.hpp>
+
+// gazebo_ros_control
+#include <gazebo_ros_control/robot_hw_sim.h>
+
+// URDF
+#include <urdf/model.h>
+
+
+
+namespace gazebo_ros_control
+{
+
+class DefaultRobotHWSim : public gazebo_ros_control::RobotHWSim
+{
+public:
+
+  virtual bool initSim(
+    const std::string& robot_namespace,
+    rclcpp::Node::SharedPtr& model_nh,
+    gazebo::physics::ModelPtr parent_model,
+    const urdf::Model *const urdf_model,
+    std::vector<transmission_interface::TransmissionInfo> transmissions);
+
+  virtual void readSim(rclcpp::Time time, rclcpp::Duration period);
+
+  virtual void writeSim(rclcpp::Time time, rclcpp::Duration period);
+
+  virtual void eStopActive(const bool active);
+
+  // Methods used to control a joint.
+  enum ControlMethod {EFFORT, POSITION, POSITION_PID, VELOCITY, VELOCITY_PID};
+protected:
+  
+  // Register the limits of the joint specified by joint_name and joint_handle. The limits are
+  // retrieved from joint_limit_nh. If urdf_model is not NULL, limits are retrieved from it also.
+  // Return the joint's type, lower position limit, upper position limit, and effort limit.
+  void registerJointLimits(const std::string& joint_name,
+                           const hardware_interface::JointStateHandle& joint_handle,
+                           const ControlMethod ctrl_method,
+                           const rclcpp::Node::SharedPtr& joint_limit_nh,
+                           const urdf::Model *const urdf_model,
+                           int *const joint_type, double *const lower_limit,
+                           double *const upper_limit, double *const effort_limit,
+                           double *const vel_limit);
+
+  unsigned int n_dof_;
+
+#if 0 //@todo
+  hardware_interface::JointStateInterface    js_interface_;
+  hardware_interface::EffortJointInterface   ej_interface_;
+  hardware_interface::PositionJointInterface pj_interface_;
+  hardware_interface::VelocityJointInterface vj_interface_;
+#endif
+#if 0 //@todo
+  joint_limits_interface::EffortJointSaturationInterface   ej_sat_interface_;
+  joint_limits_interface::EffortJointSoftLimitsInterface   ej_limits_interface_;
+  joint_limits_interface::PositionJointSaturationInterface pj_sat_interface_;
+  joint_limits_interface::PositionJointSoftLimitsInterface pj_limits_interface_;
+  joint_limits_interface::VelocityJointSaturationInterface vj_sat_interface_;
+  joint_limits_interface::VelocityJointSoftLimitsInterface vj_limits_interface_;
+#endif
+  std::vector<std::string> joint_names_;
+  std::vector<int> joint_types_;
+  std::vector<double> joint_lower_limits_;
+  std::vector<double> joint_upper_limits_;
+  std::vector<double> joint_effort_limits_;
+  std::vector<double> joint_vel_limits_;
+  std::vector<ControlMethod> joint_control_methods_;
+#if 0
+  std::vector<control_toolbox::Pid> pid_controllers_;
```

Comment:
I havent found use for control_toolbox::Pid in gazebo_ros_controls, hence I disabled it. However, in another joint_trajectory_controller draft I used control_toolbox::Pid as part the code for converting trajectories to effort values. I'll link the joint_trajectory_controller PR later.

Reply:
https://github.com/ros-controls/ros2_controllers/pull/36